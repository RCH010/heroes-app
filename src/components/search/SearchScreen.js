import React, { useMemo } from 'react'
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import HeroCard from '../heroes/HeroCard';
import queryString from 'query-string'
import { getHeroesByName } from '../../selectors/getHeroesByName';
// para trabajar con query string, existe este paquete
// https://www.npmjs.com/package/query-string
// queryString.parse(location.search)

export const SearchScreen = ({history}) => {

    const location = useLocation();
    const {q = ''} = queryString.parse(location.search);

    const [{heroSearched}, handleInputChange, reset ] = useForm({
        heroSearched: q
    });

    const filteredHeroes = useMemo(() => getHeroesByName(q), [q])

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${heroSearched}`);
    }
    return (
        <>
            <h1>Search your Hero</h1>
            <br/>
            <div className='row'>
                <div className='col-5'>
                    <h2 className='h4'>Search Form</h2>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input 
                            type='text'
                            placeholder="Your hero's name"
                            className='form-control'
                            value={heroSearched}
                            name='heroSearched'
                            autoComplete='off'
                            onChange={handleInputChange}
                        />
                        <button 
                            className='form-control btn btn-block btn-outline-primary mt-2'
                            type='submit'
                        >
                            Search
                        </button>
                    </form>
                </div>
                <div className='col-7'>
                    <h3 className='h4'>Results</h3>
                    <hr />
                    
                    {
                        (q === '') 
                            &&
                            <div className='alert alert-info'>
                                Search a Hero!
                            </div>
                    }
                    {
                        (q !== '' && filteredHeroes.length === 0) 
                            &&
                            <div className='alert alert-danger'>
                                There is no a hero with {q}
                            </div>
                    }
                    {
                        filteredHeroes.map(hero => (
                            <HeroCard 
                                key={hero.id}
                                hero={hero}
                            />
                        ))
                    }
                </div>
            </div>
        </>
    )
}
