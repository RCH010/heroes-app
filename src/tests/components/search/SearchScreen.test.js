import { mount } from 'enzyme';
import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Tests in <SearchScreen />', () => {
    
    const history = {
        push: jest.fn()
    };

    test('should display correctly with default values', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path='/search' component={() => <SearchScreen history={history} />} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a Hero!');
    })

    test('should textfield have input with value from query string', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path='/search' component={() => <SearchScreen history={history} />} />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper.find('HeroCard').exists()).toBe(true);
    })
    
    test('should display alert class if hero doesnt exists', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batmanSuperman']}>
                <Route path='/search' component={() => <SearchScreen history={history} />} />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').text().trim()).toBe('There is no a hero with batmanSuperman');
    })
    
    test('should call pushHistory when input recieved', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batmanSuperman']}>
                <Route path='/search' component={() => <SearchScreen history={history} />} />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'heroSearched',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({preventDefault(){}});
        expect(history.push).toHaveBeenCalledWith('?q=batman')
    })
    
})
