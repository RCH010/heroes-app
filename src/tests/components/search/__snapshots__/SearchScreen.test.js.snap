// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <SearchScreen /> should display correctly with default values 1`] = `
Array [
  <h1>
    Search your Hero
  </h1>,
  <br />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h2
        className="h4"
      >
        Search Form
      </h2>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="heroSearched"
          onChange={[Function]}
          placeholder="Your hero's name"
          type="text"
          value=""
        />
        <button
          className="form-control btn btn-block btn-outline-primary mt-2"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h3
        className="h4"
      >
        Results
      </h3>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero!
      </div>
    </div>
  </div>,
]
`;
